### ISSUE TITLE
Test never exits

### ISSUE THREAD

**Issue by veox** [cburgdorf/lahja#1]
(Pardon me scrapping the issue template.)

Python: CPython 3.7.0
`lahja`: commit c2f8811e0dc048255aec1f06fe17b7f525c665a9 (v0.5.0)

-----

This is to corroborate your finding in https://github.com/ethereum/py-evm/pull/1212#issuecomment-416975580, and dump some thoughts.

The test, when run as `pytest -k test_foo --capture=no` with a few tiny modifications, produces (sometimes after `CTRL+C`, sometimes while still executing the test):

``` python
task: <Task pending coro=<EventBus._start() running at /home/veox/src/lahja/lahja/eventbus.py:59> wait_for=<Future finished exception=EOFError()>>
base_events.py            1591 ERROR    Future exception was never retrieved
future: <Future finished exception=EOFError()>
Traceback (most recent call last):
  File "/usr/lib64/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/veox/src/lahja/lahja/async_util.py", line 24, in async_get
    return queue.get()
  File "/usr/lib64/python3.7/multiprocessing/queues.py", line 94, in get
    res = self._recv_bytes()
  File "/usr/lib64/python3.7/multiprocessing/connection.py", line 216, in recv_bytes
    buf = self._recv_bytes(maxlength)
  File "/usr/lib64/python3.7/multiprocessing/connection.py", line 407, in _recv_bytes
    buf = self._recv(4)
  File "/usr/lib64/python3.7/multiprocessing/connection.py", line 383, in _recv
    raise EOFError
EOFError
```

The changes to receive this is basically adding

``` python
bus.stop()

print('===== DEBUG =====') # pytest could use a newline...
print(bus._running)
print(bus._incoming_queue.qsize())
```

to the end of `test_foo()`.

What this looks to me is the `while` loop in `eventbus._start()` getting stuck on the `async_get()` function when it's already in the loop, so the semaphore-ey never gets to check `self._running` again:

https://github.com/cburgdorf/lahja/blob/c2f8811e0dc048255aec1f06fe17b7f525c665a9/lahja/eventbus.py#L53-L54

The `async_get()` uses a blocking `multiprocessing.Queue.get()`, with a decorator I don't understand:

https://github.com/cburgdorf/lahja/blob/c2f8811e0dc048255aec1f06fe17b7f525c665a9/lahja/async_util.py#L12-L24

In short, the event bus is waiting for the queue, and the queue is waiting to `recv` data.

There seems to be a lack of task cancellations, too, but I'm too inexperienced with `async` to know, leave alone give advice.

**Comment by cburgdorf** [comment]
Hey @veox thanks for looking into this. You are correct. I created a smaller test case for StackOverflow yesterday and basically it boils down to this: If you use a `ThreadPoolExecutor`, then you need to design the work that it is carrying out to be stoppable because otherwise there's no clean way to take down the executor.

https://stackoverflow.com/questions/52081033/how-to-shutdown-process-with-event-loop-and-executor#comment91115577_52081665

**Comment by cburgdorf** [comment]
@veox This should do it https://github.com/cburgdorf/lahja/commit/4d32f6b352888dd869276c495018bc106c462dee

I still sometimes see a weird internal exception flowing around that I can't make sense of but at least the tests finish reliable :sweat_smile:

**Comment by cburgdorf** [comment]
Just wanted to say that the random exceptions are gone now and it seems to work consistently. Also, the repository has moved to https://github.com/ethereum/lahja

**Comment by veox** [comment]
Cool. I did try the fix from a few hours ago, and did get the exceptions (after interpreter teardown). Haven't checked the newer one yet (doesn't look like I'll have time today).
